File: aurora-scala/src/main/scala/net/gree/aurora/scala/domain/clustergroup/ClusterGroupImpl.scala
package net.gree.aurora.scala.domain.clustergroup

import java.util.{List => JList}
import net.gree.aurora.domain.clustergroup.{ClusterGroup => JClusterGroup}
import net.gree.aurora.scala.domain.cluster.{Cluster, ClusterId}
import org.sisioh.dddbase.utils.{Try => JTry}
import scala.collection.JavaConverters._
import scala.util.{Failure, Success, Try}

private[domain]
class ClusterGroupImpl
(val underlying: JClusterGroup) extends ClusterGroup {

  val identity = ClusterGroupId(underlying.getIdentity.getValue)

  def resolveCluster(shardId: ClusterId): Try[Cluster] = {
    val result = underlying.resolveCluster(shardId)
    result match {
      case success: JTry.Success[_] =>
        Success(Cluster(success.get().asInstanceOf[Cluster]))
      case failure: JTry[_] =>
        Failure(failure.getCause)
    }
  }

  def clusters: Try[Seq[Cluster]] = {
    if (underlying.getClusters.isSuccess) {
      val clusters = underlying.getClusters.get()
      Success(clusters.asScala.map(Cluster(_)))
    } else {
      Failure(underlying.getClusters.getCause)
    }
  }

}

File: aurora-core/src/main/java/net/gree/aurora/domain/clustergroup/ClusterGroupImpl.java
package net.gree.aurora.domain.clustergroup;

import net.gree.aurora.domain.cluster.Cluster;
import net.gree.aurora.domain.cluster.ClusterId;
import net.gree.aurora.domain.cluster.ClusterRepository;
import net.gree.aurora.domain.cluster.ClusterRepositoryFactory;
import org.apache.commons.lang.Validate;
import org.sisioh.dddbase.model.impl.AbstractEntity;
import org.sisioh.dddbase.utils.Try;

import java.util.List;

final class ClusterGroupImpl
        extends AbstractEntity<ClusterGroupId>
        implements ClusterGroup {

    private final Long order;
    private final ClusterRepository shardRepository = ClusterRepositoryFactory.create();

    /**
     * インスタンスを生成する。
     *
     * @param identity {@link org.sisioh.dddbase.model.Identity}
     * @param clusters {@link Cluster}のリスト
     * @param order    オーダー
     */
    public ClusterGroupImpl(ClusterGroupId identity, List<Cluster> clusters, Long order) {
        super(identity);
        Validate.notNull(clusters);
        Validate.notNull(order);
        for (Cluster shard : clusters) {
            shardRepository.store(shard);
        }
        this.order = order;
    }

    public Try<Cluster> resolveCluster(ClusterId shardId) {
        return shardRepository.resolve(shardId);
    }

    @Override
    public Try<List<Cluster>> getClusters() {
        return Try.ofSuccess(shardRepository.toList());
    }

    @Override
    public Long getOrder() {
        return order;
    }

    @Override
    public int compareTo(ClusterGroup o) {
        return order.compareTo(o.getOrder());
    }

    @Override
    public String toString() {
        return String.format("ClusterGroup(identity = %s, clusters = %s)", getIdentity(), getClusters());
    }

    @Override
    public ClusterGroupImpl clone() {
        return (ClusterGroupImpl) super.clone();
    }
}
